equinox/nn/_weight_norm.py

```diff
@@ -0,0 +1,107 @@
+from typing import Generic, Optional, TypeVar, Union
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array, PRNGKeyArray
+
+from .._module import field, Module
+from .._tree import tree_at
+
+
+_Layer = TypeVar("_Layer")
+
+
+def _norm_except_axis(
+    v: Array, pow: Optional[Union[int, str]] = None, axis: Optional[int] = 0
+) -> Array:
+    norm_func = lambda x: jnp.linalg.norm(x, ord=pow, keepdims=True)
+    vmapped_norm_func = lambda axis: jax.vmap(
+        norm_func,
+        in_axes=axis,
+        out_axes=axis,
+    )(v)
+    if axis is not None:
+        return vmapped_norm_func(axis)
+    else:
+        return norm_func(vmapped_norm_func(0)).reshape([])
```

Comment:
Did you have something like this in mind @patrick-kidger? I couldn't figure out how to vmap over all axes, so I opted for this solution. I guess this could also be achieved as a scan? Though I'm not sure what would be gained by this.

Reply:
So I think something like this might be what you're after:
```python
from jaxtyping import Array, PyTree, Scalar

def _norm_except_axis(v: Array, norm: Callable[[PyTree[Array]], Scalar], axis: Optional[int]):
    if axis is None:
        return norm(x)
    else:
        return jax.vmap(norm, in_axes=axis, out_axes=axis)(x)
```
where for bonus points I've also (a) removed default arguments, and (b) made the norm selectable (so that one can pass in any norm function -- we have a number of these defined elsewhere in the Equinox ecosystem, we could imagine consolidiating them here). In practice right now we can call this with `norm=lambda x: jnp.linalg.norm(x, keepdims=True)`.

WDYT?

Reply:
This looks great! Somehow I thought that passing `None` to `jnp.linalg.norm` doesn't work - not sure why.

One minor addition, we'll have to reshape the `if axis is None` branch if we want the same behaviour as the torch equivalent. We've already strayed from it a little by having `axis=None` denote computing the norm over all dimensions as opposed to `-1` (though I think that is the better approach here, as the torch API is confusing and not consistent, as `-2` for example works like indexing, i.e. computes the norm over the second to last axis), so I guess we could further abandon a strict adherence here.

Additionally, if we're now making this function a bit more general purpose, it might make sense to move it elsewhere and add dedicated tests for it, what do you think? Do you have a good place for this in mind? Nothing immediately jumps out at me scrolling through the codebase.

Do we want to add defaulted arguments also?

So taken together it would look like this:

```python
from functools import partial

def _norm_except_axis(v: Array
, norm: Callable[[PyTree[Array]], Scalar] = partial(jnp.linalg.norm, keepdims=True)
, axis: Optional[int] = 0):
    if axis is None:
        return norm(v).reshape([])
    else:
        return jax.vmap(norm, in_axes=axis, out_axes=axis)(v)
```

Reply:
Yup, I think departing from the PyTorch API makes sense here.

I think I'm happy to leave the `_norm_except_axis` function here. It's pretty small, we can worry about that later if we ever need it. :)

I definitely wouldn't add default arguments. It's a general good practice for internal APIs not to include default arguments, as these can be a source of silently surprising behaviour. (Especially here, where we are always passing every argument explicitly anyway!)

Reply:
Good point regarding the defaulted arguments, I agree.

Leaving the function where it is then for now then :) 

I've made all the changes discussed - additionally, I've imported from `beartype.typing` as opposed to `typing` directly to silence a PEP 585 related warning (see [here](https://beartype.readthedocs.io/en/latest/api_roar/#pep-585-deprecations)). I hope this is OK? It looks like `equinox` has a  dependency on`beartype` already (correct me if I'm wrong!)

Reply:
Equinox should only use `beartype` for its tests -- we don't have it as a dependency outside of that.

So in our case the appropriate thing to do is:
```python
from collections.abc import Callable
from typing import Generic, Optional, TypeVar
```

(A lot of things in the `typing` module have been deprecated: some in favour of a builtin, e.g. `List -> list`, and some in favour of an item of `collections.abc`, e.g. `Callable`. Fixing up these warnings is usually a matter of making one of those two changes.)

Reply:
import `Callable` from `collections.abc` fixed the warning, thank you! `beartype` should suggest this as a fix in their documentation. Might make a PR for it.